_posts/2023-02-06-qml-01-data-encoding.md

```diff
@@ -0,0 +1,403 @@
+
+```
+# Baidu Quantum Paddle
+import paddle
+from paddle_quantum.ansatz import Circuit
+from paddle_quantum.gate import BasisEncoding, AmplitudeEncoding, AngleEncoding
+import paddle_quantum as pq
+
+# Xanadu PennyLane
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.templates.embeddings import BasisEmbedding, AmplitudeEmbedding  # Importa o template
+
+# IBM Qiskit
+from qiskit import Aer, QuantumCircuit, execute
+
+
+# Auxiliares
+import math
+```
+
+### 2. Basis Encoding
+
+Primeiramente coletamos nossos dados númericos clássicos e o transformamos em cadeias de bits clássicos. Daí, aproveitamos a string de bits de entrada e o relacionamos com o estado quântico através de uma transformação bem intuitiva:
+
+$$
+010011101 \rightarrow \ket{010011101}
+$$
+
+Isto é, para cada input, usaremos um estado da base computacional. Nosso espaço de estados quânticos, agora denotado por $$ \ket{H} $$ e será uma combinação linear de todos os estados transformados:
+
+$$
+\ket{H} = \frac{1}{\sqrt{M}} \sum_{m=1}^{M} \ket{x^m}
+$$
+
+Exemplo:
+Seja $$ H = \left\{x^1=101, x^2=111 \right\} $$. Daí:
+
+$$
+\ket{H} = \frac{1}{\sqrt{2}} ( \ket{101} + \ket{111})
+$$
+
+#### 2.1 Qiskit
+
+```
+n=16       #Nesse caso, número de qubits = tamanho do nosso string binário
+qc = QuantumCircuit(n)
+x = '1011100010101110'   #Bit clássico que queremos transformar em quântico
+x=x[::-1]    #Inverte a string (só é necessário porque o padrão é inverter os qubits)
+
+# Adicionamos um gate X (not) em cada componente do circuito para criar o qubit desejado
+for i in range(len(x)):
+    if x[i] == '1':
+        qc.x(i)
+  
+qc.draw()
+
+backend = Aer.get_backend('statevector_simulator')
+result = execute(qc,backend).result().get_statevector()
+result.draw('latex')
+```
+
+Podemos também utilizar o método decompose() do Qiskit para sabermos exatamente quais gates são necessários para chegarmos nesse estado quântico.
+
+#### 2.2 Paddle
+
+```
+
+# Seguindo o método apresentado na seção anterior:
+n = 4
+basis_enc = Circuit(n)    # Inicializa nosso circuito com o Paddle Quantum
+x = '1011'
+
+for i in range(len(x)):
+    if x[i] == '1':
+        basis_enc.x(i)
+  
+print(basis_enc)
+
+
+# Para execução do circuito usamos:
+init_state = pq.state.zero_state(n)
+basis_quantum_state = basis_enc(init_state)
+print(basis_quantum_state)
+
+
+# Entretanto o Paddle Quantum já tem uma função pronta para Basis Encoding:
+built_in_basis_enc = BasisEncoding(num_qubits=n)
+x = paddle.to_tensor([1, 0, 1, 1])   # Mudamos a informação clássica x para o tipo Tensor
+built_in_basis_enc_state = built_in_basis_enc(feature=x)
+
+print(built_in_basis_enc_state)
+```
+
+
+#### 2.3 PennyLane
+
+Em [ref.V] é usado um código parecido com o abaixo para implementação.
+
+```
+dev = qml.device('default.qubit', wires=6)
+
+@qml.qnode(dev)
+def circuit(data):
+    for i in range(6):
```

Comment:
In this case, I think you could use the data length to the first for loop as you have been doing in the others. Don't you agree?

Reply:
Concordo. Fica mais geral. Obrigadoo